README.md
# Blue Rectangle

A 2D demo platformer game made using [Dunai](https://github.com/ivanperez-keera/dunai), [Bearriver](https://hackage.haskell.org/package/bearriver) and [Haskell SDL2 bindings](https://hackage.haskell.org/package/sdl2). This game implements a variant of continuous collision detection.

Rules:
- You are the Blue Rectangle.
- Red Rectangles are the enemies.
- To complete a level eliminate all of the Red Rectangles by jumping on top of them.
- Level is restarted if you collide with an enemy not by jumping on top of it.


Controls:
- A to move left
- D to move right
- Space to jump

![Gameplay](./assets/gameplay.gif)

The game implements continuous collision detection following an idea suggested in the paper [Functional Reactive Programming, Refactored (Perez et al.)](https://www.cs.nott.ac.uk/~psxip1/papers/2016-HaskellSymposium-Perez-Barenz-Nilsson-FRPRefactored-short.pdf). Section 7.1.3 outlines how predicted collision times can be calculated from within the monadic stream functions and how the times can be retrieved from the top level reactimate functions. This is done by using `CCDGameMonad` which is defined in the paper and which allows holding predicted collision times in writer monad context. Example given in the paper is a simple case of a bouncing ball in one dimension. In this game the idea is implemented in 2D setting with multiple moving objects.

For reference, bellow is the definition of `CCDGameMonad` used in this project, `tellFututureTime` is an MSF that puts a `FutureTime` value in the writer context (the context holds the minimal told value).

```haskell
type DTime = Double

data FutureTime = AtTime DTime | Infinity
  deriving (Eq, Ord, Show)

instance Semigroup FutureTime where
  (<>) = min

instance Monoid FutureTime where
  mempty :: FutureTime
  mempty = Infinity

type CCDGameMonad m = ReaderT DTime (WriterT FutureTime m)

type GameMonad = (CCDGameMonad Identity)

tellFutureTime :: MSF GameMonad FutureTime ()
tellFutureTime = liftTransS $ arrM tell
```


## Rough outline of the game physics

Each game object (enemy, player, static environment block) holds its trajectory data (position, velocity, acceleration). Accelerations are constant between the collision events and control changes. So the trajectories can be thought to be piecewise parabolas. At every step:

- For each object its trajectory is update (according to the current parabola), and new control information is gathered.
- Taking all of the object trajectories, new collision information is calculated.
- Taking collision information and control information of all the objects, new trajectories are calculated and switching event is produced to update object trajectories.
- Based on new trajectories, nearest collision time is predicted (since trajectories are parabolas, it boils down to solving a bunch of quadratic equations).

At the top level game logic loop, the nearest collision time can be extracted from monadic context. If needed, multiple physics steps can be performed for every rendering sample, taking into account extracted predicted collision times and not demanding physics steps with times beyond nearest collision (to avoid bullet effect).

## Running the game (Linux)

1. Install GHC and cabal. One way to do this is using [GHCup](https://www.haskell.org/ghcup/install/).


2. Install SDL2 library dependencies:
    ```
    sudo apt-get install libsdl2-dev libsdl2-image-dev libsdl2-ttf-dev libsdl2-mixer-dev
    ```
3. Clone the repository.

4. Build and run the game using cabal inside the root folder of the repository:
    ```
    cabal run
    ```

## Relevant literature

I would like to point to some literature that I reviewed while working on this project and from which I took some ideas: 

- [Antony Courtney, Henrik Nilsson, John Peterson. The Yampa Arcade.](https://dl.acm.org/doi/10.1145/871895.871897) A good extended example on how to structure games with heterogeneous actors using Yampa library.
- [Ivan Perez, Manuel Bärenz,  Henrik Nilsson. Functional Reactive Programming, Refactored.](https://www.cs.nott.ac.uk/~psxip1/papers/2016-HaskellSymposium-Perez-Barenz-Nilsson-FRPRefactored-short.pdf) An introduction to monadic stream functions (MSFs) and Dunai library. It explains how MSFs can be used to implement different flavours of FRP and support implementing a several different game mechanics.
- [Ivan Perez. Extensible and Robust Functional Reactive Programming.](http://www.cs.nott.ac.uk/~psxip1/papers/2017-Perez-thesis-latest.pdf) Parts of this PhD thesis covers topics on applications of monadic stream functions in game programming in greater detail, it could be used as a supplement to [FRP Refactored](https://www.cs.nott.ac.uk/~psxip1/papers/2016-HaskellSymposium-Perez-Barenz-Nilsson-FRPRefactored-short.pdf) paper.


CHANGELOG.md
# Revision history for blue-rectangle

## 0.1.0.0 -- 2023-09-21

* First version. Released on an unsuspecting world.


app/Types.hs
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}

module Types where

import DPSwitch (dpSwitch)
import IdentityList

import Control.Monad.Trans.MSF
import Data.Functor.Identity
import Data.MonadicStreamFunction
import FRP.BearRiver (Event (Event, NoEvent))

import qualified Data.Text as T

import qualified Data.Bifunctor
import Data.Text (Text)

type DTime = Double

data FutureTime = AtTime DTime | Infinity
  deriving (Eq, Ord, Show)

instance Semigroup FutureTime where
  (<>) = min

instance Monoid FutureTime where
  mempty :: FutureTime
  mempty = Infinity

type CCDGameMonad m = ReaderT DTime (WriterT FutureTime m)

type GameMonad = (CCDGameMonad Identity)

data Rct = Rct
  { pX :: Double -- x coord of the bottom left corner
  , pY :: Double -- y coord of the bottom left corner
  , w :: Double
  , h :: Double
  }
  deriving (Eq)

data Vct = Vct
  { x :: Double
  , y :: Double
  }
  deriving (Eq)

type Object = MSF GameMonad ObjInput ObjOutput

data ObjInput = ObjInput
  { oiOtherActors :: IL ObjOutput
  , oiControlKeys :: ControlKeys
  , oiHit :: FRP.BearRiver.Event ()
  , oiBounce :: FRP.BearRiver.Event ()
  }

data ObjOutput = ObjOutput
  { ooObsObjState :: ObsObjState
  , ooRemoveReq :: FRP.BearRiver.Event ()
  , ooControlIntention :: ControlIntention
  , ooControlIntentionChanged :: Bool
  }

data GameOutput = GameOutput
  { goObjOutputs :: IL ObjInput
  , goText :: [Text]
  }

data MovementState = MovementState
  { msRct :: Rct
  , msVel :: Vct
  , msAcc :: Vct
  }
  deriving (Eq)

data BlockingData = BlockingData
  { up :: Bool
  , right :: Bool
  , down :: Bool
  , left :: Bool
  , upRight :: Bool
  , downRight :: Bool
  , downLeft :: Bool
  , upLeft :: Bool
  }
  deriving (Eq)

blockingDataDefault :: BlockingData
blockingDataDefault = BlockingData True True True True True True True True

data ControlKeys = ControlKeys
  { keyW :: Bool
  , keyA :: Bool
  , keyS :: Bool
  , keyD :: Bool
  , keyUp :: Bool
  , keyLeft :: Bool
  , keyDown :: Bool
  , keyRight :: Bool
  , keyEnter :: Bool
  , keyEscape :: Bool
  , keySpace :: Bool
  }
  deriving (Eq)

data CtrlScheme = CtrlScheme
  { cUp :: ControlKeys -> Bool
  , cLeft :: ControlKeys -> Bool
  , cDown :: ControlKeys -> Bool
  , cRight :: ControlKeys -> Bool
  }

letterScheme :: CtrlScheme
letterScheme = CtrlScheme keyW keyA keyS keyD

arrowScheme :: CtrlScheme
arrowScheme = CtrlScheme keyUp keyLeft keyDown keyRight

letterSpaceScheme :: CtrlScheme
letterSpaceScheme = CtrlScheme keySpace keyA keyS keyD

controlKeysDefault :: ControlKeys
controlKeysDefault =
  ControlKeys
    { keyW = False
    , keyA = False
    , keyS = False
    , keyD = False
    , keyUp = False
    , keyLeft = False
    , keyDown = False
    , keyRight = False
    , keyEnter = False
    , keyEscape = False
    , keySpace = False
    }

data Direction = DirLeft | DirRight | DirNone
  deriving (Eq)

data ControlIntention = ControlIntention
  { ciDir :: Direction
  , ciJump :: Bool
  , ciBounce :: Bool
  , ciControlKeys :: ControlKeys -- perhaps redundant
  , ciSideVel :: Double
  , ciJumpVel :: Double
  , ciBounceVel :: Double
  , ciGravity :: Double
  }

defaultControlIntention :: ControlIntention
defaultControlIntention =
  ControlIntention
    { ciDir = DirNone
    , ciJump = False
    , ciBounce = False
    , ciControlKeys = controlKeysDefault
    , ciBounceVel = 0
    , ciJumpVel = 0
    , ciGravity = 0
    , ciSideVel = 0
    }

data ObsObjState
  = OOSPlayer
      { oosMovementState :: MovementState
      , oosBlockingData :: BlockingData
      , ctrlScheme :: CtrlScheme
      }
  | OOSEnemy
      { oosMovementState :: MovementState
      , oosBlockingData :: BlockingData
      , oosLeftBound :: Maybe Double
      , oosRightBound :: Maybe Double
      }
  | OOSStaticBlock
      { oosMovementState :: MovementState
      }

isPlayer :: ObsObjState -> Bool
isPlayer (OOSPlayer{}) = True
isPlayer _ = False
isEnemy :: ObsObjState -> Bool
isEnemy (OOSEnemy{}) = True
isEnemy _ = False
isStaticBlock :: ObsObjState -> Bool
isStaticBlock (OOSStaticBlock{}) = True
isStaticBlock _ = False


app/Levels.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}

module Levels where

import GameCore
import IdentityList
import Types

boundaries :: [Rct]
boundaries =
  [ Rct 0 0 800 30
  , Rct 0 0 30 1000
  , Rct 770 0 30 1000
  ]

map1 :: [Rct]
map1 =
  boundaries
    ++ [Rct 200 100 50 50]

map2 :: [Rct]
map2 =
  boundaries
    ++ [ Rct 100 100 300 50
       , Rct 450 200 300 50
       , Rct 200 400 250 50
       , Rct 500 320 50 50
       ]

map3 :: [Rct]
map3 =
  boundaries
    ++ [ Rct 100 100 50 25
       , Rct 100 300 50 25
       , -- , Rct 100 500  50  10
         Rct 200 200 50 25
       , Rct 200 400 50 25
       , Rct 300 400 50 25
       , Rct 450 200 100 300
       -- , Rct 200 600  50  10
       ]

map4 :: [Rct]
map4 =
  boundaries
    ++ [ Rct 200 30 50 70
       ]

platformWithAnEnemy :: Rct -> Double -> Double -> Double -> [Object]
platformWithAnEnemy platformRct xOffset enemyWidth enemyHeight =
  let platform = staticObject $ MovementState platformRct (Vct 0 0) (Vct 0 0)
      enemy =
        enemyObject
          ( MovementState
              ( Rct
                  (pX platformRct + xOffset)
                  (pY platformRct + h platformRct + 1)
                  enemyWidth
                  enemyHeight
              )
              (Vct 0 0)
              (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , Just $ pX platformRct
          , Just $ pX platformRct + w platformRct - enemyWidth
          )
   in [platform, enemy]

level1 :: IL Object
level1 =
  listToIL
    ( [ playerObject
          ( MovementState (Rct 100 50 20 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , letterSpaceScheme
          )
      ]
        ++ platformWithAnEnemy (Rct 300 200 340 20) 0 50 50
        ++ platformWithAnEnemy (Rct 500 300 240 20) 20 50 50
        ++ map (\rct -> staticObject $ MovementState rct (Vct 0 0) (Vct 0 0)) map1
    )

level2 :: IL Object
level2 =
  listToIL
    ( [ enemyObject
          ( MovementState (Rct 100 150 20 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , Just 100
          , Just 380
          )
      , enemyObject
          ( MovementState (Rct 600 250 20 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , Just 450
          , Just 730
          )
      , enemyObject
          ( MovementState (Rct 300 500 50 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , Just 200
          , Just 400
          )
      , playerObject
          ( MovementState (Rct 100 50 20 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , letterSpaceScheme
          )
      ]
        ++ map (\rct -> staticObject $ MovementState rct (Vct 0 0) (Vct 0 0)) map2
    )

level3 :: IL Object
level3 =
  listToIL
    ( [ enemyObject
          ( MovementState (Rct 500 40 20 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , Nothing
          , Nothing
          )
      , playerObject
          ( MovementState (Rct 100 40 20 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , letterSpaceScheme
          )
      ]
        ++ platformWithAnEnemy (Rct 300 200 150 20) 10 40 30
        ++ platformWithAnEnemy (Rct 640 250 200 20) 10 50 50
        ++ map (\rct -> staticObject $ MovementState rct (Vct 0 0) (Vct 0 0)) map3
    )

level4 :: IL Object
level4 =
  listToIL
    ( [ enemyObject
          ( MovementState (Rct 500 40 20 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , Nothing
          , Nothing
          )
      , playerObject
          ( MovementState (Rct 100 40 20 50) (Vct 0 0) (Vct 0 0)
          , blockingDataDefault
          , defaultControlIntention
          , letterSpaceScheme
          )
      ]
        ++ platformWithAnEnemy (Rct 280 100 600 20) 10 50 50
        ++ platformWithAnEnemy (Rct 550 200 200 20) 15 60 40
        ++ platformWithAnEnemy (Rct 450 330 200 20) 40 40 40
        ++ platformWithAnEnemy (Rct 200 300 200 20) 80 60 40
        ++ platformWithAnEnemy (Rct 150 400 100 20) 50 40 40
        ++ map (\rct -> staticObject $ MovementState rct (Vct 0 0) (Vct 0 0)) map4
    )

multiLevelSpec =
  multiLevelGame'
    [
      ( level1
      ,
        [ "Level 1."
        , "You are the Blue Rectangle."
        , "A and D to move, Space to jump."
        , "Eliminate the Red Rectangles!"
        ]
      )
    , (level2, ["Level 2.", "Eliminate the Red Rectangles!"])
    , (level3, ["Level 3.", "Eliminate more of the Red Rectangles!"])
    , (level4, ["Level 4.", "Eliminate even more of the Red Rectangles!"])
    ]


app/DPSwitch.hs
{-# LANGUAGE Arrows #-}
{-# LANGUAGE ExplicitForAll #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}

module DPSwitch where

import qualified Control.Arrow as Bifunctor
import Data.Foldable (sequenceA_, traverse_)
import Data.Interval
import Data.Maybe (fromMaybe)
import Data.MonadicStreamFunction.InternalCore
import Data.Traversable as T
import FRP.BearRiver (
  Event (Event, NoEvent),
  SF,
  boolToEvent,
  dSwitch,
  dpSwitchB,
  dup,
  edge,
  iEdge,
  integral,
  lMerge,
  loopPre,
  mergeBy,
  notYet,
  parB,
  parC,
  rMerge,
  switch,
  tag,
  time,
 )

-- dpSwitch is a part of Yampa but it has not been ported in any official Dunai release.
-- But there exists a pull request implementing dpSwitch here:
-- https://github.com/ivanperez-keera/dunai/pull/316
-- I took that implementation and adapted it

dpSwitch ::
  (Monad m, Traversable col) =>
  (forall sf. (a -> col sf -> col (b, sf))) ->
  col (FRP.BearRiver.SF m b c) ->
  SF m (a, col c) (FRP.BearRiver.Event d) ->
  (col (SF m b c) -> d -> SF m a (col c)) ->
  SF m a (col c)
dpSwitch rf sfs sfF sfCs = MSF $ \a -> do
  let bsfs = rf a sfs
  res <- T.mapM (\(b, sf) -> unMSF sf b) bsfs
  let cs = fmap fst res
      sfs' = fmap snd res
  (e, sfF') <- unMSF sfF (a, cs)
  let ct = case e of
        FRP.BearRiver.Event d -> sfCs sfs' d
        NoEvent -> dpSwitch rf sfs' sfF' sfCs
  return (cs, ct)


app/IdentityList.hs
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}

-- Adding Identity List according to this paper:
-- https://www.researchgate.net/publication/228785589_The_Yampa_Arcade

module IdentityList where

type ILKey = Int
data IL a = IL
  { ilNextKey :: ILKey
  , ilAssocs :: [(ILKey, a)]
  }

emptyIL :: IL a
emptyIL = IL 0 []

insertIL_ :: a -> IL a -> IL a
insertIL_ elem ilist@(IL nextKey assocs) = IL (nextKey + 1) ((nextKey, elem) : assocs)

listToIL :: [a] -> IL a
listToIL = foldr insertIL_ emptyIL

elemsIL :: IL a -> [a]
elemsIL ilist@(IL _ assocs) = map snd assocs

assocsIL :: IL a -> [(ILKey, a)]
assocsIL ilist@(IL _ assocs) = assocs

deleteIL :: ILKey -> IL a -> IL a
deleteIL key ilist@(IL nextKey assocs) = IL nextKey (filter (\assoc -> fst assoc /= key) assocs)

replaceIL :: ILKey -> a -> IL a -> IL a
replaceIL key newVal ilist@(IL nextKey assocs) = IL nextKey (map (\(k, val) -> (k, if key == k then newVal else val)) assocs)

mapIL :: ((ILKey, a) -> b) -> IL a -> IL b
mapIL f ilist@(IL nextKey assocs) = IL nextKey (map (\(key, elem) -> (key, f (key, elem))) assocs)

instance Functor IL where
  fmap f = mapIL (f . snd)

instance Foldable IL where
  foldr f init ilist@(IL _ assocs) = foldr f init (fmap snd assocs)

instance Traversable IL where
  traverse :: (Applicative f) => (a -> f b) -> IL a -> f (IL b)
  traverse f arr = listToIL <$> traverse (f . snd) (assocsIL arr)


app/CollisionUtils.hs
{-# LANGUAGE Arrows #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}

module CollisionUtils where

import DPSwitch (dpSwitch)
import IdentityList
import Types

import Control.Monad.Trans.MSF
import Data.Functor.Identity
import Data.MonadicStreamFunction
import Data.MonadicStreamFunction.InternalCore (MSF (MSF, unMSF))
import FRP.BearRiver (parC)

import Data.IORef
import qualified Data.Text as T

import qualified Data.Bifunctor
import Data.Text (Text)

epsilon :: Double
epsilon = 0.01

-- MSF to put times until collision to continuous collision detection monad context
tellFutureTime :: MSF GameMonad FutureTime ()
tellFutureTime = liftTransS $ arrM tell

getImpactTimeWithAcc :: Double -> Double -> Double -> FutureTime
getImpactTimeWithAcc p v a
  | a == 0 = getImpactTime p v
  | abs p <= (epsilon / 2) = Infinity
  | otherwise = impactTimeWithAcc
 where
  discriminant
    | v < 0 = v * v - 2 * a * (p - (epsilon / 4))
    | otherwise = v * v - 2 * a * (p + (epsilon / 4))
  root1 = (-(sqrt discriminant) - v) / a
  root2 = (sqrt discriminant - v) / a
  impactTimeWithAcc
    | discriminant < 0 = Infinity
    | root1 > 0 && root2 > 0 = min (AtTime root1) (AtTime root2)
    | root1 > 0 || root2 > 0 = max (AtTime root1) (AtTime root2)
    | otherwise = Infinity

-- Needed to play with epsilons as there were some issues with double arithmethic and rounding
getImpactTime :: Double -> Double -> FutureTime
getImpactTime p v
  | p > epsilon / 2 && v < 0 = AtTime (-(p - (epsilon / 3)) / v)
  | p < -epsilon / 2 && v > 0 = AtTime (-(p + (epsilon / 3)) / v)
  | p > 0 && v < 0 = Infinity -- AtTime (-p/v)
  | p < 0 && v > 0 = Infinity -- AtTime (-p/v)
  | otherwise = Infinity

-- This function tells the time of an intersection between any of the infinitely extended parallel edges of two rectangles
-- (e.g. time when x coordinates of left sides of two rectangles intersect,
-- time when x coordinate of left edge of one box and right edge of another box intersect, etc.)
-- This can be optimised, so that only times of actual intersections of two rectangles are reported.
-- The only issue right now is that physics simulations might be told to be run more frequently than necessary,
-- which impacts performance
tellHittingTimeToBox :: MSF GameMonad (MovementState, MovementState) ()
tellHittingTimeToBox = proc (ms1@(MovementState box1 vel1 acc1), ms2@(MovementState pos2 vel2 acc2)) -> do
  tellFutureTime
    -<
      getImpactTimeWithAcc
        (pX pos2 + w pos2 - pX box1)
        (x vel2 - x vel1)
        (-x acc2 + x acc1)
  tellFutureTime
    -<
      getImpactTimeWithAcc
        (-pX pos2 + w box1 + pX box1)
        (-x vel2 + x vel1)
        (-x acc2 + x acc1)
  tellFutureTime
    -<
      getImpactTimeWithAcc
        (pY pos2 + h pos2 - pY box1)
        (y vel2 - y vel1)
        (-y acc2 + y acc1)
  tellFutureTime
    -<
      getImpactTimeWithAcc
        (-pY pos2 + h box1 + pY box1)
        (-y vel2 + y vel1)
        (-y acc2 + y acc1)

calcTimeToHit :: MSF GameMonad ([MovementState], MovementState) ()
calcTimeToHit = proc (trips, trip) -> do
  parC tellHittingTimeToBox -< map (\t -> (t, trip)) trips
  returnA -< ()

speedWhenGrounded :: Bool -> Bool -> Double -> Double
speedWhenGrounded blockedLow blockedHigh y
  | y < 0 && blockedLow = 0
  | y > 0 && blockedHigh = 0
  | otherwise = y

checkAABB :: Rct -> Rct -> Bool
checkAABB box1 box2 =
  (pX box1 + w box1) > pX box2
    && pX box1 < (pX box2 + w box2)
    && (pY box1 + h box1) > pY box2
    && pY box1 < (pY box2 + h box2)

getBlockingData :: Rct -> [Rct] -> BlockingData
getBlockingData (Rct x y w h) rcts =
  let u = foldr ((||) . checkAABB (Rct x (y + epsilon) w h)) False rcts
      r = foldr ((||) . checkAABB (Rct (x + epsilon) y w h)) False rcts
      d = foldr ((||) . checkAABB (Rct x (y - epsilon) w h)) False rcts
      l = foldr ((||) . checkAABB (Rct (x - epsilon) y w h)) False rcts
      ur = foldr ((||) . checkAABB (Rct (x + epsilon) (y + epsilon) w h)) False rcts
      dr = foldr ((||) . checkAABB (Rct (x + epsilon) (y - epsilon) w h)) False rcts
      dl = foldr ((||) . checkAABB (Rct (x - epsilon) (y - epsilon) w h)) False rcts
      ul = foldr ((||) . checkAABB (Rct (x - epsilon) (y + epsilon) w h)) False rcts
   in BlockingData
        u
        r
        d
        l
        ur
        dr
        dl
        ul


app/Main.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}

module Main where

import DPSwitch (dpSwitch)
import GameCore
import IdentityList
import Levels
import Types

import Control.Concurrent (threadDelay)
import Control.Monad (guard, when)
import Control.Monad.Trans.MSF (
  ReaderT (runReaderT),
  WriterT (runWriterT),
 )
import qualified Data.Bifunctor
import Data.Functor.Identity (Identity (runIdentity))
import Data.IORef (IORef, readIORef, writeIORef)
import Data.MonadicStreamFunction (MSF)
import Data.MonadicStreamFunction.InternalCore (MSF (MSF, unMSF))
import Data.Text (Text)
import qualified Data.Text as T
import Foreign.C.Types (CInt)
import SDL
import qualified SDL
import SDL.Font (Font, initialize, load, solid)
import SDL.Raw (Rect (Rect), quitRequested, vkLoadLibrary)
import qualified SDL.Raw.Font as SDL.Font
import SDL.Vect ()

screenWidth, screenHeight :: CInt
(screenWidth, screenHeight) = (800, 640)

frameStep :: Double
frameStep = 0.02

microSecsPerSec :: Double
microSecsPerSec = 1e6

main :: IO ()
main = do
  SDL.initialize [SDL.InitVideo]
  SDL.Font.initialize

  window <-
    SDL.createWindow
      "Blue Rectangle"
      SDL.defaultWindow{SDL.windowInitialSize = V2 screenWidth screenHeight}
  SDL.showWindow window

  -- This font was selected
  -- https://github.com/theleagueof/orbitron
  font <- SDL.Font.load "assets/font.ttf" 30

  renderer <-
    SDL.createRenderer
      window
      (-1)
      SDL.RendererConfig
        { SDL.rendererType = SDL.AcceleratedRenderer
        , SDL.rendererTargetTexture = False
        }
  screenSurface <- SDL.getWindowSurface window

  SDL.updateWindowSurface window

  runGame renderer window font multiLevelSpec 0
  quitGame window

quitGame :: Window -> IO ()
quitGame window = do
  SDL.destroyWindow window
  SDL.quit

handleSDLEvent :: ControlKeys -> SDL.Event -> ControlKeys
handleSDLEvent ctrl ev =
  case eventPayload ev of
    KeyboardEvent (KeyboardEventData _ _ _ (Keysym ScancodeA _ _)) -> ctrl{keyA = not (keyA ctrl)}
    KeyboardEvent (KeyboardEventData _ _ _ (Keysym ScancodeD _ _)) -> ctrl{keyD = not (keyD ctrl)}
    KeyboardEvent (KeyboardEventData _ _ _ (Keysym ScancodeS _ _)) -> ctrl{keyS = not (keyS ctrl)}
    KeyboardEvent (KeyboardEventData _ _ _ (Keysym ScancodeW _ _)) -> ctrl{keyW = not (keyW ctrl)}
    KeyboardEvent (KeyboardEventData _ _ _ (Keysym ScancodeUp _ _)) -> ctrl{keyUp = not (keyUp ctrl)}
    KeyboardEvent (KeyboardEventData _ _ _ (Keysym ScancodeLeft _ _)) -> ctrl{keyLeft = not (keyLeft ctrl)}
    KeyboardEvent (KeyboardEventData _ _ _ (Keysym ScancodeDown _ _)) -> ctrl{keyS = not (keyDown ctrl)}
    KeyboardEvent (KeyboardEventData _ _ _ (Keysym ScancodeRight _ _)) -> ctrl{keyW = not (keyRight ctrl)}

processInput :: Bool -> IO (ControlKeys, Bool)
processInput _ = do
  quitReq <- quitRequested
  es <- pollEvents
  keyHeld <- getKeyboardState
  let ctrl =
        ControlKeys
          { keyA = keyHeld ScancodeA
          , keyW = keyHeld ScancodeW
          , keyS = keyHeld ScancodeS
          , keyD = keyHeld ScancodeD
          , keyUp = keyHeld ScancodeUp
          , keyLeft = keyHeld ScancodeLeft
          , keyDown = keyHeld ScancodeDown
          , keyRight = keyHeld ScancodeRight
          , keyEnter = keyHeld ScancodeKPEnter
          , keyEscape = keyHeld ScancodeEscape
          , keySpace = keyHeld ScancodeSpace
          }
  pure (ctrl, quitReq)

nextTimeStep :: FutureTime -> Double -> DTime
nextTimeStep Infinity delayTime = delayTime -- frameStep
nextTimeStep (AtTime time) delayTime = min time delayTime -- frameStep

-- Run single step of a MSF and extracts future time (from the writer context in GameMonad)
-- and continuation of MSF for next iterations
singleIteration ::
  MSF GameMonad ControlKeys (IL ObjOutput, [Text]) ->
  ControlKeys ->
  Double ->
  Identity
    ( ((IL ObjOutput, [Text]), MSF GameMonad ControlKeys (IL ObjOutput, [Text]))
    , FutureTime
    )
singleIteration myMSF ctrl t = do
  (out, futureTime) <- runWriterT $ runReaderT (unMSF myMSF ctrl) t
  return (out, futureTime)

-- Take several samples of the game,
-- each time extracting the next calculated future time
-- and decrementing the remaining time
runIterationsForGivenTime ::
  ControlKeys ->
  MSF GameMonad ControlKeys (IL ObjOutput, [Text]) ->
  DTime ->
  DTime ->
  IO
    ( ((IL ObjOutput, [Text]), MSF GameMonad ControlKeys (IL ObjOutput, [Text]))
    , FutureTime
    )
runIterationsForGivenTime ctrl myMSF firstStepTime 0 = do
  return $ runIdentity $ singleIteration myMSF ctrl firstStepTime
runIterationsForGivenTime ctrl myMSF firstStepTime furtherStepsTime = do
  let ((objOut, msf), newFutureTime) = runIdentity $ singleIteration myMSF ctrl firstStepTime
  let (AtTime nextStepTime) = min (AtTime furtherStepsTime) newFutureTime
  let remainingTime = furtherStepsTime - nextStepTime
  runIterationsForGivenTime ctrl msf nextStepTime remainingTime

-- Using simple clock technique
-- 1. Getting time before and after game logic step
-- 2. Running game logic step for fixed amount of time (this does not need to be fixed)
-- 3. Rendering
-- 4. Getting the time at the end of iteration
-- 5. Sleeping for the difference between fixed iteration step and measured time differences (with some lower bound on sleeping time)
-- TODO: restructure this function into something more similar to reactimate loop
runGame ::
  Renderer ->
  Window ->
  Font ->
  MSF GameMonad ControlKeys (IL ObjOutput, [Text]) ->
  DTime ->
  IO ()
runGame rend wind font myMSF t = do
  loopStartTime <- SDL.time
  input@(ctrl, quitReq) <- processInput True

  if keyEscape ctrl || quitReq
    then return ()
    else do
      ((outs, msf), ft) <- runIterationsForGivenTime ctrl myMSF t (frameStep - t)

      renderScene outs rend wind font

      loopEndTime <- SDL.time
      let delayTime = max (frameStep - (loopEndTime - loopStartTime)) 0 -- (frameStep / 1000)
      threadDelay $ fromIntegral $ floor $ delayTime * microSecsPerSec

      runGame rend wind font msf (nextTimeStep ft frameStep)

renderScene :: (IL ObjOutput, [Text]) -> Renderer -> Window -> Font -> IO ()
renderScene (objOut, textArray) rend wind font = do
  SDL.rendererDrawColor rend $= V4 20 30 20 maxBound
  clear rend

  mapM_ (renderObject rend) (snd <$> assocsIL objOut)
  renderText textArray 10 font wind rend
  SDL.present rend

renderText :: [Text] -> CInt -> Font -> SDL.Window -> SDL.Renderer -> IO ()
renderText [] _ _ _ _ = return ()
renderText (x : xs) yOffset font wind rend = do
  text <- SDL.Font.solid font (V4 200 200 200 maxBound) x
  dims@(V2 xDim yDim) <- SDL.surfaceDimensions text
  textTexture <- SDL.createTextureFromSurface rend text
  SDL.freeSurface text
  SDL.copy rend textTexture Nothing (Just $ Rectangle (P $ V2 40 yOffset) dims)
  SDL.destroyTexture textTexture
  renderText xs (yOffset + yDim + 5) font wind rend

renderObject :: Renderer -> ObjOutput -> IO ()
renderObject rend obj = do
  let xc = pX $ msRct $ oosMovementState $ ooObsObjState obj
  let yc = pY $ msRct $ oosMovementState $ ooObsObjState obj
  let wc = fromIntegral $ floor $ w $ msRct $ oosMovementState $ ooObsObjState obj
  let hc = fromIntegral $ floor $ h $ msRct $ oosMovementState $ ooObsObjState obj
  colourAction (ooObsObjState obj)
  SDL.fillRect
    rend
    ( Just $
        SDL.Rectangle
          ( P $ V2 (fromIntegral $ floor xc) (fromIntegral $ -hc + screenHeight - floor yc)
          )
          (V2 wc hc)
    )
 where
  colourAction :: ObsObjState -> IO ()
  colourAction OOSPlayer{} = SDL.rendererDrawColor rend $= V4 20 20 150 maxBound
  colourAction OOSStaticBlock{} = SDL.rendererDrawColor rend $= V4 20 200 200 maxBound
  colourAction OOSEnemy{} = SDL.rendererDrawColor rend $= V4 200 20 20 maxBound
